### ISSUE TITLE
tota11y as an extension, instead of a bookmarklet

### ISSUE THREAD

**Issue by jdan** [Khan/tota11y#62]
Since starting tota11y I've been on the fence to make it a full-fledged browser extension, or keep it as a bookmarklet.

**Bookmarklet pros:**
* Easy to install
* Easy to support multiple browsers
* Script is portable - teams can include it in the dev-environment (we do this at Khan Academy)

**Bookmarklet cons:**
* Requires manual activation (click the bookmarklet) on every pageload
* Can't save preferences across sites
* **Can't be loaded on every site** (tota11y injects a `<script>` tag into the DOM because its code is so large, sites like GitHub block these requests)

**Extension pros:**
* Can be loaded on every site
* Can be loaded automatically on every pageload
* Can have cross-site options (toggle which plugins you want)
* Access to an extension API - which we'll need to audit things like unsemantic buttons (`getEventListeners(DOMElement)`)

**Extension cons:**
* If your company uses tota11y, each employee would have to install it manually (this is one of the reasons I originally opted for a bookmarklet)
* Must maintain multiple browser releases

There's also the question of having *both* a bookmarklet and extension.

Would love to hear your thoughts!

**Comment by NiklasBr** [comment]
:+1: for an extension, I don't have my bookmarks bar visible in my browsers (to get more room for content) and have to memorise the keyboard shortcuts for them. I get them wrong about 10-20% of the time. A problem totally of my own making :) but I like it like that and would prefer the extension.

**Comment by jgardella** [comment]
I say either do just the extension, or do both. The cons are too big to only offer tota11y as a bookmarklet. Especially given that the bookmarklet just won't work at all on some sites.

**Comment by seriema** [comment]
Me and a friend created one but are unsure how it should be connected with the official repo. It could either be a separate repo and just inject the script as the bookmarklet does, which is what we did now. Or we could somehow merge this so that every time tota11y publishes a new version a new Chrome version is created automatically and bundles the script. This later choice would allow offline usage and maybe toggling tota11y on/off or other features.

So currently we have some of the bookmarklet cons but PR's are welcome!

Here's the [repo](https://github.com/kvasir/tota11y-chrome) and [Chrome web store](https://chrome.google.com/webstore/detail/tota11y-for-chrome/nkghaekndgmonifcpfgjmpfjlhnmflhp).

**Comment by wavebeem** [comment]
This would be really rad! I think more people would be likely to use this if it was in the toolbar instead of hidden in bookmarks.

**Comment by tricinel** [comment]
I've created a Firefox add-on as well. I've used the Chrome extension and missed it when I changed browsers. The bookmarklet worked just fine, but I never have my bookmarks toolbar visible...

This one bundles the script so it does work offline, but every new release of tota11y would require me to update the add-on (which I don't mind at all).

Grab it [here](https://addons.mozilla.org/en-US/firefox/addon/tota11y-accessibility-toolkit/)

**Comment by cruxicheiros** [comment]
Has anyone considered making a Tampermonkey userscript for this? It would be browser-agnostic (as TamperMonkey is available on all major browsers, and a couple less popular ones) and would most likely solve the same set of issues.

**Comment by Fetz** [comment]
I would vote to do both 👍.

Maybe could be a mono repo (maybe this? https://github.com/lerna/lerna).

In a project a worked, I started to create a bookmarklet and then I build a simple chrome extension that its only purpose was to inject the bookmarklet in the page.

**Comment by jameswilson** [comment]
In addition to the one from @seriema I just did a search and found two other Tota11y Chrome Extensions. Because I wasn't sure which one to pick, here is a complete list of the current offerings:

Extension | Author | Version | Last updated | Size
---------- | ------- | ------- | ------------- | ----
[tota11y for Chrome](https://chrome.google.com/webstore/detail/tota11y-for-chrome/nkghaekndgmonifcpfgjmpfjlhnmflhp) | @seriema | v1.0.0 | 2016-02-14 | 34.25KiB
[Tota11y Plugin, from Khan Academy](https://chrome.google.com/webstore/detail/tota11y-plugin-from-khan/oedofneiplgibimfkccchnimiadcmhpe) | John Hartley | v0.1.5 | 2016-06-29 | 194KiB
[tota11y chrome extension](https://chrome.google.com/webstore/detail/tota11y-chrome-extension/jbhkjcigeionejpngkcdccblocdnjini) | @kevinforrestconnors | v1.61 |  2015-07-10 | 171KiB

It looks like the one from @seriema is the only one that doesn't bundle the script.

**Comment by kevinforrestconnors** [comment]
@jameswilson yeah, I haven't updated mine in forever, and last time I checked there were better options available, so I didn't feel the need to update.  Do you think I should remove it from the store so that people will have an easier time finding the better extensions?

**Comment by jameswilson** [comment]
@kevinforrestconnors there are a lot of accessibility extensions these days, and it is really hard to know which one to use without trial and error.  I guess removing yours would probably help narrow the choice down, particularly if the version of the tota11y library that is bundled is outdated.

**Comment by nickytonline** [comment]
I hadn't checked the issues and thought it would be fun to make this an extension, and then the first issue I see is this one. 🙃 I forked the repo and just added some assets, a manifest and made a very basic Chrome extension that gets bundled via webpack. I threw up a PR, #131, if you want to take a look. Although there is a maintenance burden, I think making it an extension makes the tool look more legit. Since this is just a POC, I only created a Chrome extension.

**Comment by AndrewCraswell** [comment]
For what it's worth, Microsoft for a number of years struggled with not having enough a11y tools to reduce the friction in testing for accessibility issues. One of the things that made the work difficult, was that in addition to WCAG and Section 508 standards, Microsoft had a more strict superset of these guidelines, the Microsoft Accessibility Standards (MAS). Building tooling around the MAS is much harder because most existing projects work around the WCAG or Section 508 standards, so much that was done for Microsoft needed to be custom. More recently, they are doing away with MAS and building standardized tooling around more open specifications so the tools can be open sourced and make use of community feedback and contributions.

To this end, there used to be a internal only Chrome extension for MAS testing, which has been revamped and released publicly as [Accessibility Insights](https://accessibilityinsights.io/). From what I remember, this tool (at least when I used it) addressed some of the pros mentioned in this issue, including cross-site options, projects configured for site testing with bug logging, etc. Could be useful for ideas, or contributions in areas that tota11y excels.

For reference, the GitHub code lives [here](https://github.com/Microsoft/accessibility-insights-web).

**Comment by kevinforrestconnors** [comment]
> @kevinforrestconnors there are a lot of accessibility extensions these days, and it is really hard to know which one to use without trial and error. I guess removing yours would probably help narrow the choice down, particularly if the version of the tota11y library that is bundled is outdated.

Because mine is outdated, and to aid in people finding useful extensions, i've removed mine from the store.
